gramex/__init__.py

```diff
@@ -291,3 +292,51 @@ def shutdown():
     if ioloop_running(ioloop):
         app_log.info('Shutting down Gramex...')
         ioloop.stop()
+
+
+def log(handler=None, **kwargs):
+    '''
+    Logs structured information for future reference. Typical usage::
+
+        gramex.log(level='INFO', x=1, msg='abc')
+
+    This logs ``{level: INFO, x: 1, msg: abc, port: 9988, time: 2020-07-21 13:41:00}``. 3 keys are
+    added:
+
+    1. ``level``: logging level. Defaults to INFO
+    2. ``time``: current time as YYYY-MM-DD HH:MM:ZZ in UTC
+    3. ``port``: application's current port
+
+    If a logging service like ElasticSearch has been configured, it will periodically flush the
+    logs into ElasticSearch.
+
+    By default, this logs into the ``default`` index. If you have multiple log indices, you can
+    optionally add an ``_index="log-name"``. For example, if the ``gramex.yaml`` configuration was
+    like this::
+
+        gramexlog:
+            app1:
+                host: localhost
+                index: app1
+            app2:
+                host: localhost
+                index: app2
+
+    ... then, ``gramex.log(x=1, y=2, _index='app1')`` will log into the ``app1`` index.
+    '''
+    # If any argument is a
+    if handler is not None:
+        kwargs.update({key: val[0] for key, val in handler.args.items()})
+    from . import services
+    conf = services.info.gramexlog
+    if conf and 'queue' in conf and 'maxlength' in conf:
+        kwargs['level'] = kwargs.get('level', 'INFO').upper()
+        kwargs['time'] = datetime.datetime.utcnow().strftime('%Y-%m-%d %H:%M:%SZ')
+        kwargs['port'] = app_log_extra['port']
+        idx = kwargs.get('_index', 'default')
+        kwargs['_index'] = conf['conf'].get(idx).get('index')
```

Comment:
Let's handle these situations:

- `log(_index='nonexistent')` -- raise an Exception. No need to handle
- The `index:` key is not specified in the configuration. Then, default it to the app name
- Allow multiple applications

So this code can go into `log_to_es()`
```python
        idx = kwargs.get('_index', 'default')
        kwargs['_index'] = conf['conf'][idx].get('index', idx)
```